qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -410,43 +414,122 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
+                node.op = control_flow.map_blocks(self._run_inner, node.op)
+                new_dag.apply_operation_back(node.op, node.qargs, node.cargs)
 
-            qubits = (
-                [dag.find_bit(x).index for x in node.qargs] if self._use_qubit_indices else None
-            )
+            elif getattr(node.op, "_directive", False):
+                new_dag.apply_operation_back(node.op, node.qargs, node.cargs)
 
-            decomposition, modified = self._recursively_handle_op(node.op, qubits)
+            elif dag.has_calibration_for(node) or len(node.qargs) < self._min_qubits:
+                new_dag.apply_operation_back(node.op, node.qargs, node.cargs)
 
-            if not modified:
-                continue
+            else:
+                node_qubits = (
+                    [dag.find_bit(x).index for x in node.qargs] if self._use_qubit_indices else None
+                )
 
-            if isinstance(decomposition, QuantumCircuit):
-                dag.substitute_node_with_dag(
-                    node, circuit_to_dag(decomposition, copy_operations=False)
+                if use_ancillas:
+                    clean_ancillas_available = [q for q in clean_ancillas if q not in node.qargs]
+                    dirty_ancillas_available = [
+                        q
+                        for q in dag.qubits
+                        if q not in node.qargs and q not in clean_ancillas_available
+                    ]
+                else:
+                    clean_ancillas_available = []
+                    dirty_ancillas_available = []
+
+                decomposition, modified = self._recursively_handle_op(
+                    node.op,
+                    node_qubits,
+                    num_clean_ancillas=len(clean_ancillas_available),
+                    num_dirty_ancillas=len(dirty_ancillas_available),
                 )
-            elif isinstance(decomposition, DAGCircuit):
-                dag.substitute_node_with_dag(node, decomposition)
-            elif isinstance(decomposition, Operation):
-                dag.substitute_node(node, decomposition)
 
-        return dag
+                if not modified:
+                    new_dag.apply_operation_back(node.op, node.qargs, node.cargs)
+
+                else:
+                    if isinstance(decomposition, QuantumCircuit):
+                        decomposition = circuit_to_dag(decomposition, copy_operations=False)
+
+                    if isinstance(decomposition, DAGCircuit):
+                        # Compute the qubits on which to apply the decomposition
+                        # (including auxiliary qubits).
+                        num_ancillas_used = decomposition.num_qubits() - len(node.qargs)
+                        if not use_ancillas or num_ancillas_used == 0:
+                            extended_qargs = node.qargs
+                        elif num_ancillas_used <= len(clean_ancillas_available):
```

Comment:
It might be good to state this behavior somewhere: that clean ancillas, if available, are preferably used over dirty ones.